02-peer-protocol.md

```diff
@@ -1713,6 +1784,186 @@ satoshis, which is possible if `dust_limit_satoshis` is below 546 satoshis).
 
+`OP_RETURN` is only standard if followed by PUSH opcodes, and the total script
+is 83 bytes or less. We are slightly stricter, to only allow a single PUSH, but
+there are two forms in script: one which pushes up to 75 bytes, and a longer
+one (`OP_PUSHDATA1`) which is needed for 76-80 bytes.
+
+### Closing Negotiation: `closing_complete` and `closing_sig`
+
+Once shutdown is complete, the channel is empty of HTLCs, there are no commitments
+for which a revocation is owed, and all updates are included on both commitments,
+the final current commitment transactions will have no HTLCs.
+
+Each peer creates their own closing transaction where they pay the fee, and sends
+`closing_complete` to the other peer with the transaction details. The other peer
+simply signs that transaction and sends back `closing_sig`. Each peer will thus
+independently send `closing_complete` and receive `closing_sig`, resulting in two
+independent (but conflicting) closing transactions being created.
+
+The lesser-paid peer (if either is) can opt to omit their own output from the
+closing transaction.
+
+This process can be repeated multiple times by sending `closing_complete` again,
+which allows increasing the fees and changing the output script.
+
+1. type: 40 (`closing_complete`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u16`:`closer_scriptpubkey_len`]
+   * [`closer_scriptpubkey_len*byte`:`closer_scriptpubkey`]
+   * [`u16`:`closee_scriptpubkey_len`]
+   * [`closee_scriptpubkey_len*byte`:`closee_scriptpubkey`]
+   * [`u64`:`fee_satoshis`]
+   * [`u32`:`locktime`]
+   * [`closing_tlvs`:`tlvs`]
+
+1. type: 41 (`closing_sig`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u16`:`closer_scriptpubkey_len`]
+   * [`closer_scriptpubkey_len*byte`:`closer_scriptpubkey`]
+   * [`u16`:`closee_scriptpubkey_len`]
+   * [`closee_scriptpubkey_len*byte`:`closee_scriptpubkey`]
+   * [`u64`:`fee_satoshis`]
+   * [`u32`:`locktime`]
+   * [`closing_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `closing_tlvs`
+2. types:
+    1. type: 1 (`closer_output_only`)
+    2. data:
+        * [`signature`:`sig`]
+    1. type: 2 (`closee_output_only`)
+    2. data:
+        * [`signature`:`sig`]
+    1. type: 3 (`closer_and_closee_outputs`)
+    2. data:
+        * [`signature`:`sig`]
+
+#### Requirements
+
+Note: the details and requirements for the transaction being signed are in [BOLT 3](03-transactions.md#closing-transaction).
+
+An output is *dust* if the amount is less than the [Bitcoin Core Dust Thresholds](03-transactions.md#dust-limits).
+
+Both nodes:
+  - After a `shutdown` has been sent and received, AND no HTLCs remain in either commitment transaction:
+    - SHOULD send a `closing_complete` message.
+
+The sender of `closing_complete` (aka. "the closer"):
+  - MUST set `fee_satoshis` to a fee less than or equal to its outstanding balance, rounded down to whole satoshis.
+  - MUST set `fee_satoshis` so that at least one output is not dust.
```

Comment:
I still find this requirement a bit weird...since I can only influence my local output amount, and following this strictly means I need to choose a fee such that my local output is not dust. Meanwhile we allow it to be dust, as long as `OP_RETURN` is used. And by using `OP_RETURN` it means there's nothing to do with the fees I set.